src/AuthenticationService.php

```diff
@@ -428,4 +433,72 @@ public function getLoginRedirect(ServerRequestInterface $request): ?string
 
         return $parsed['path'] . $parsed['query'];
     }
+
+    /**
+     * Impersonates a user
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @param \Psr\Http\Message\ResponseInterface $response The response
+     * @param \ArrayAccess $impersonator User who impersonates
+     * @param \ArrayAccess $impersonated User impersonated
+     * @return array
+     */
+    public function impersonate(
+        ServerRequestInterface $request,
+        ResponseInterface $response,
+        \ArrayAccess $impersonator,
+        \ArrayAccess $impersonated
+    ): array {
+        /** @var \Authentication\Authenticator\PersistenceInterface $provider */
+        $provider = $this->getAuthenticationProvider();
+        if (!($provider instanceof ImpersonationInterface)) {
+            $className = get_class($provider);
+            throw new \InvalidArgumentException(
+                "The {$className} Provider must implement ImpersonationInterface in order to use impersonation."
+            );
+        }
+
+        return $provider->impersonate($request, $response, $impersonator, $impersonated);
+    }
+
+    /**
+     * Stops impersonation
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @param \Psr\Http\Message\ResponseInterface $response The response
+     * @return array
+     */
+    public function stopImpersonating(ServerRequestInterface $request, ResponseInterface $response): array
+    {
+        /** @var \Authentication\Authenticator\PersistenceInterface $provider */
+        $provider = $this->getAuthenticationProvider();
+        if (!($provider instanceof ImpersonationInterface)) {
+            $className = get_class($provider);
+            throw new \InvalidArgumentException(
+                "The {$className} Provider must implement ImpersonationInterface in order to use impersonation."
+            );
+        }
+
+        return $provider->stopImpersonating($request, $response);
+    }
+
+    /**
+     * Returns true if impersonation is being done
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @return bool
+     */
+    public function isImpersonating(ServerRequestInterface $request): bool
+    {
+        /** @var \Authentication\Authenticator\ImpersonationInterface $provider */
+        $provider = $this->getAuthenticationProvider();
```

Comment:
The code on 494 - 500 is repeated thrice. They could be moved into a separate protected method `getImpersonationProvider()`.

Reply:
Fixed